### ISSUE TITLE
DioError's response is always null inside the interceptor's onError 

### ISSUE THREAD

**Issue by ahmedmehanna** [flutterchina/dio#942]
### New Issue Checklist

- [x] I have searched for a similar issue in the [project](https://github.com/flutterchina/dio/issues) and found none

### Issue Info

| Info                           | Value                                                 |
| ------------------------------ | ----------------------------------------------------- |
| Platform Name                  | flutter / ios / android                          |
| Platform Version               | 1.20.3                             |
| Dio Version                    | 3.0.10                                  |
| VS Code Version | 1.49.1              |
| Repro rate                     | all the time (100%)  |

### Issue Description and Steps
DioError's response is always null inside the interceptor's onError override function

```
class InternalServerErrorInterceptor extends Interceptor {
  @override
  onError(DioError error) async {
    if (error.response != null) {       // <--- the response is always null
      if (error.response.statusCode >= 500 && error.response.statusCode < 600) {
        return InternalServerError();
      }
    }
    return null;
  }
}
```
```
class ApiClient {
  ApiClient._() {
    _apiDio = new Dio();
    _apiDio.options.baseUrl = Constants.apiBaseUrl;
    _apiDio.options.connectTimeout = Duration(minutes: 3).inMilliseconds;
    _apiDio.options.receiveTimeout = Duration(minutes: 3).inMilliseconds;
    _apiDio.interceptors.add(AuthInterceptor());
    _apiDio.interceptors.add(InternalServerErrorInterceptor());
    _apiDio.interceptors.add(UnauthorizedInterceptor());
    _apiDio.interceptors.add(BadRequestInterceptor());
  }

  static final ApiClient instance = ApiClient._();

  static Dio _apiDio;

  post(String path, dynamic data, Function badRequestToModelError, {Options options}) async {
    try {
      return await _apiDio.post(
        path,
        data: data,
        options: options,
      );
    } on BadRequestError catch (error) {
      throw badRequestToModelError(error);
    } on UnauthenticatedError {
      throw UnauthenticatedError();
    } catch (error) {
      throw ApiError();
    }
  }
}
```

**Comment by marzoukali** [comment]
I have the same issue. I even downgraded Dio to the previous version but it didn't work .. @dio team can you please help??

**Comment by marzoukali** [comment]
Just adding the interceptor before setting the duration, fix it for me:

```
    _apiDio.options.baseUrl = Constants.apiBaseUrl;
    _apiDio.interceptors.add(BadRequestInterceptor())
    _apiDio.options.connectTimeout = Duration(minutes: 3).inMilliseconds;
    _apiDio.options.receiveTimeout = Duration(minutes: 3).inMilliseconds;
```

**Comment by marzoukali** [comment]
Just make sure that you call the super.onError(err) and not passing null if there is nothing to intercept. Interceptors are a chain so if you passing null that will terminate the response with exactly the same issue you posted.

```
class BadRequestInterceptor extends Interceptor {
  @override
  onError(DioError error) async {
    if (error.response != null) {
      if (error.response.statusCode == 400) {
        throw BadRequestError(error.response.data);
      } else if (error.response.statusCode == 417) {
        return BadRequestError({});
      }
    }
    return super.onError(error);
  }
}
```

**Comment by ahmedmehanna** [comment]
Thanks, This solution fixed the issue on Dio v3.0.10
